### PR TITLE
build: add conventional commits

### LINKED ISSUE #344: Enforce Conventional Commits
The following is commented away in our README but we never implemented this feature. It is a great way to standardize commit messages as well as to produce a great standardized input for automation tools, which for example, can automatically produce change logs for every deployment.

> 1. We [automatically enforce](https://github.com/conventional-changelog/commitlint) the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format for commit messages. This can be frustrating for junior developers, but I promise that you'll quickly get used to it! -->

> The Conventional Commits specification is a lightweight convention on top of commit messages. It provides an easy set of rules for creating an explicit commit history; which makes it easier to write automated tools on top of. -->

### PR DESCRIPTION
**Pull request by rndquu**
Resolves #344

This PR enables check for commit messages.

There can be 2 flows:
1. When a new user git clones the project and runs `yarn install` then commit msg hook is installed in this `postinstall` script https://github.com/ubiquity/ubiquity-dollar/compare/development...rndquu:issue/344?expand=1#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R35
2. Existing users (who have already git cloned the project) have to manually run `yarn postinstall` in order to install commit msg hooks

Before this PR linter was not running in pre-commit hook, but now pre-commit hook is enabled and linter is running. I commented linter code because:
1. it is not clear what is the best strategy to run the linter
2. reenabling linter is out of scope of this task

P.S. I closed the previous PR https://github.com/ubiquity/ubiquity-dollar/pull/353 because it has conflicts in `yarn.lock` and after solving the conflicts the build fails due to https://github.com/ubiquity/ubiquity-dollar/issues/377

### CODE REVIEW

**Comment by 0x4007** [comment]
I wish there was a better solution for 2. Any ideas on how to enforce this? Perhaps we can on CI? Just make sure that the git hook is automatically installed after the developer runs `yarn` to install dependencies. That's quite common to do after a pull.

**Comment by zgorizzo69** [comment]
@rndquu anyway people can still bypass it if they want so for me it is fine we just have to check that commits are ok from time to time

**Comment by 0x4007** [comment]
> @rndquu anyway people can still bypass it if they want so for me it is fine we just have to check that commits are ok from time to time

I believe that we could fork a script somewhere (instead of handwriting a new one) to enforce in CI

File: .husky/pre-commit
Comment (0x4007) [inline .husky/pre-commit]: I don't understand what you mean by "best way to lint" 

It runs on changed files using `lint-staged`

It should load the settings we already defined.

File: package.json
Comment (0x4007) [inline package.json:36]: I assume that this is needed for `lint-staged` to run?

File: .husky/pre-commit
Comment (rndquu) [reply .husky/pre-commit]: `lint-staged` is enabled again

I meant 2 things:
1. enabling linter is out of scope for this task
2. I personally don't like when code is updated via automatic tools like `prettier --write` or especially `eslint --fix`. By "best way to lint" I meant relaxing linting rules as much as possible.

File: package.json
Comment (rndquu) [reply package.json:36]: No, script command is run from https://github.com/ubiquity/ubiquity-dollar/blob/8f295b7eb0983254b26b5f3dab9f480a88831b96/.husky/pre-commit

I tried to remove husky config from `package.json` and everything worked fined so I decided not to keep commands in 2 places: `package.json` and `.husky` folder thus leaving only `.husky` folder.

**Comment by rndquu** [comment]
@pavlovcik @zgorizzo69 

udpates:
1. linter is enabled
2. added github action for checking commit msgs

So now if user skips msg check on commit via `--no-verify` then CI will block the PR

File: .github/workflows/enforce-conventional-commits.yml
Comment (0x4007) [inline .github/workflows/enforce-conventional-commits.yml:5]:
```suggestion
    branches: * 
```

I'm not sure if this is the correct syntax but we should enforce on every branch because what if a bounty hunter has a main branch for their PR and then sub branches with the wrong commit styles. Wouldn't those commits be difficult to parse for any automated tooling down the line?

**Review by 0x4007 — CHANGES_REQUESTED** [review]

**Comment by 0x4007** [comment]
I guess its hard for me to verify if the CI works because CI only runs from `development` 

I suppose we could always merge and then finish on `development` if needed.

What happens if the bounty hunter has a ton of non-compliant commits previously? Does CI only check the latest one? If they have to rename all of their previous commits that could be terrible but I can't think of a better solution.

File: .github/workflows/enforce-conventional-commits.yml
Comment (rndquu) [reply .github/workflows/enforce-conventional-commits.yml:5]: Currently the merge flow is the following:
1. Bounty hunters create PRs to the `development` branch
2. You manually merge `development` into `main` from time to time

There is no way to miss invalid commit msg because all PRs are getting in the `development` branch first

**Comment by rndquu** [comment]
> I guess its hard for me to verify if the CI works because CI only runs from development

CI action runs on PR to the `development` branch. So you pushed a commit https://github.com/ubiquity/ubiquity-dollar/pull/378/commits/6f0e4d2116d651af6122bc48c18feb01b35e5733 which expectedly failed the CI action because of invalid commit msg name

> Does CI only check the latest one? 

it runs on all commit messages

> What happens if the bounty hunter has a ton of non-compliant commits previously?

This will be hard to solve. Bounty hunter will have to squash all PR commits into a single one with a valid commit msg, smth like `git reset --soft HEAD~3 && git commit -m "feat: new feature"`

**Comment by 0x4007** [comment]
> This will be hard to solve. Bounty hunter will have to squash all PR commits into a single one with a valid commit msg, smth like `git reset --soft HEAD~3 && git commit -m "feat: new feature"`

I have a feeling that this will be an issue in the future, but again I can't figure out a better solution for this 😰

**Comment by 0x4007** [comment]
Can you demonstrate what you need to do here?

I hope that there is a way to fix this issue without deleting previous commits! PR reviews are much easier without deletes/force-pushes

I enabled auto-merge so assuming that your next commit passes, you will be able to merge this in to complete the bounty!

File: .github/workflows/enforce-conventional-commits.yml
Comment (pavlovcik) [reply .github/workflows/enforce-conventional-commits.yml:5]: I would much rather have the bounty hunters handle the squash merge / fix commit message issue than me!

File: package.json
Comment (0x4007) [reply package.json:36]: Awesome work!

**Review by 0x4007 — APPROVED** [review]

**Comment by rndquu** [comment]
@pavlovcik 

FYI, this PR is somehow automerged even with failed CI check

1. There were 2 CI issues: unresolved conversations (https://github.com/ubiquity/ubiquity-dollar/pull/378#discussion_r1045203487) and failed CI action "Conventional Commits" because of your test commit with invalid message (https://github.com/ubiquity/ubiquity-dollar/pull/378/commits/6f0e4d2116d651af6122bc48c18feb01b35e5733)
2. Then I clicked "resolve conversation" and PR automatically merged even with failed CI action

This is a pretty weird behaviour of github or I just don't understand smth about "auto merges" 

Anyway the PR was ready for merging

File: .husky/pre-commit
Comment (0x4007) [reply .husky/pre-commit]: Looks like you disabled it again.

**Comment by 0x4007** [comment]
Looks like this became a bit of a mess. 

You disabled the prettier again on the latest commit in the pre-commit hook, which should remain enabled. It's still unclear to me why you want to remove it. I think that linter/prettier should not be loosely enforced because then we will accrue formatting issues over time. Since the tools are automated, it shouldn't be burdensome for the developers working on the code either.

It looks like the CI didn't run on your latest commit so I presume that it only checks on the merge PR event instead of the commit event (like when you write a GitHub Action and how it is supposed to run "on pull_request" or "on commit") which could be after your CI is supposed to run or something. I am unsure.

**Comment by rndquu** [comment]
> You disabled the prettier again on the latest commit in the pre-commit hook, which should remain enabled

Where do you see it? This line https://github.com/ubiquity/ubiquity-dollar/blob/development/.husky/pre-commit#L4 is uncommented + I've just checked the latest build from `development` branch and `lint-staged` is running as expected on commit.

**Comment by 0x4007** [comment]
I realized I was looking at your first commit, not your latest - sorry!

**Pull request merged**
